### DISCUSSION TITLE
How to make Pocketbase work with server side rendering - SSR

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#989] [Q&A] [answered]
Is it possible to be able to use Golang to display content that is also duplicated with Svelte?  I have an app (that I'm not working on now) that is storing some Quill editor created article bodies in base64.  If you hit the page then Golang will serve it with html/template.  Then if you click on another article it will be refreshed using Svelte.  It's pretty easy to duplicate the HTML fragments because I'm using classless HTML.

Is this possible using Pocketbase?  I understand that it uses bearer tokens that must be included in the header of every request.  Is there any kind of fancy way to make the browser automatically send the headers for a hard page load?  Or perhaps because I'm extending/embedding Pocketbase there could be a way to duplicate typical sessions.

Perhaps another way of asking the question is: Is it possible to embed/extend Pocketbase so that I can completely ignore Javascript for signin, and hard page reloads?  Can I use Pocketbase with zero javascript?  Oh, and one caveat.  The routes need to be protected by auth :).

**Comment by ganigeorgiev** [comment, marked as answer]
This is hard to hard to answer without a code sample of what you are trying to do.

PocketBase is stateless. In general it needs the `Authorization` header token so that it can register its related auth model into the current request context.

To some extend you can have a session-like behavior by adding the auth token as a cookie (make sure that at least `Secure` and `SameSite=Strict` are set). Then you can register a custom echo middleware that will read that cookie and insert the related auth model into the request `echo.Contex`, something like:

```go
package main

import (
	"log"

	"github.com/labstack/echo/v5"
	"github.com/pocketbase/pocketbase"
	"github.com/pocketbase/pocketbase/apis"
	"github.com/pocketbase/pocketbase/core"
	"github.com/pocketbase/pocketbase/tokens"
	"github.com/pocketbase/pocketbase/tools/security"
	"github.com/spf13/cast"
)

func loadAuthContextFromCookie(app core.App) echo.MiddlewareFunc {
	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			tokenCookie, err := c.Request().Cookie("token")
			if err != nil || tokenCookie.Value == "" {
				return next(c) // no token cookie
			}

			token := tokenCookie.Value

			claims, _ := security.ParseUnverifiedJWT(token)
			tokenType := cast.ToString(claims["type"])

			switch tokenType {
			case tokens.TypeAdmin:
				admin, err := app.Dao().FindAdminByToken(
					token,
					app.Settings().AdminAuthToken.Secret,
				)
				if err == nil && admin != nil {
					// "authenticate" the admin
					c.Set(apis.ContextAdminKey, admin)
				}
			case tokens.TypeAuthRecord:
				record, err := app.Dao().FindAuthRecordByToken(
					token,
					app.Settings().RecordAuthToken.Secret,
				)
				if err == nil && record != nil {
					// "authenticate" the app user
					c.Set(apis.ContextAuthRecordKey, record)
				}
			}

			return next(c)
		}
	}
}

func main() {
	app := pocketbase.New()

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		e.Router.Use(loadAuthContextFromCookie(app))

		return nil
	})

	if err := app.Start(); err != nil {
		log.Fatal(err)
	}
}
```

The above will load the authenticated admin/user in the request context, which will allow you to use the existing `apis` middlewares to restrict access to your routes, eg. `apis.RequireRecordAuth()`.

**Reply by dschissler** [reply]
@ganigeorgiev That seems good enough to be included in the middlewares packages and added to a documentation.  Also it would be good to have a standard and most secure way to send this cookie to the user.  Also an included way to clear the cookie when signing out.  I understand that currently I'll have to do all of this.  Maybe you could put this in an "extras" repository.

**Reply by dschissler** [reply]
Just to make it easier for anyone just starting out:

```go
	app.OnRecordAuthRequest().Add(func(e *core.RecordAuthEvent) error {
		e.HttpContext.SetCookie(&http.Cookie{
			Name:     middleware.AuthCookieName,
			Value:    e.Token,
			Secure:   true,
			SameSite: http.SameSiteStrictMode,
			HttpOnly: true,
			MaxAge:   int(app.Settings().RecordAuthToken.Duration),
			Path:   "/",
		})
		return nil
	})

	app.OnBeforeServe().Add(func(e *core.ServeEvent) error {
		r := e.Router

		r.Use(middleware.AuthFromCookie(app))
		return nil
	})
```

edit: I had to add `Path:   "/",` to allow the cookie to actually work on my own routes.

```go
package middleware

const AuthCookieName = "token"

func AuthFromCookie(app core.App) echo.MiddlewareFunc {
	return func(next echo.HandlerFunc) echo.HandlerFunc {
		return func(c echo.Context) error {
			tokenCookie, err := c.Request().Cookie(AuthCookieName)
			if err != nil || tokenCookie.Value == "" {
				// no token cookie
				return next(c)
			}

			token := tokenCookie.Value

			claims, err := security.ParseUnverifiedJWT(token)
			if err != nil {
				return next(c)
			}
			tokenType := cast.ToString(claims["type"])

			switch tokenType {
			case tokens.TypeAdmin:
				admin, err := app.Dao().FindAdminByToken(
					token,
					app.Settings().AdminAuthToken.Secret,
				)
				if err == nil && admin != nil {
					fmt.Println("Setting admin by cookie")
					// "authenticate" the admin
					c.Set(apis.ContextAdminKey, admin)
				}
			case tokens.TypeAuthRecord:
				record, err := app.Dao().FindAuthRecordByToken(
					token,
					app.Settings().RecordAuthToken.Secret,
				)
				if err == nil && record != nil {
					fmt.Println("Setting user by cookie")
					// "authenticate" the app user
					c.Set(apis.ContextAuthRecordKey, record)
				}
			}

			return next(c)
		}
	}
}
```

**Reply by dschissler** [reply]
@ganigeorgiev I'm not quite sure yet, but I'm thinking that it might be a bad idea to combine the admin and user token into one for cookies.  It would probably be better to have like `admin-token` and `user-token`.  The reason is that it's stored in LocalStorage as `pb_admin_auth` and `pocketbase_auth`, but then that is squashed down to just a single variable for the cookie.

In your experience is that a bad thing?  Are you actually using that middleware example somewhere like that, or is it purely an example?

So I'm not deep into it yet, but it seems better to separate them and then to then at least have the option of the security middleware and verifiers (in route) to be able to have a crack at it.  One thing is if you wanted to have your UI treat everything as a normal user action even if an admin is also signed in.  So you could have an "unlock" feature, or have it in a separate UI that uses different middleware.

I plan to post some more examples of this as I get used to it.

**Reply by dschissler** [reply]
I also made some helpers like this below.  I feel like that the context access stuff should be wrapped since it's such a common operation.  Does the extra function call make a real difference in performance?

```go
func User(c echo.Context) (*models.Record, error) {
	record, _ := c.Get(apis.ContextAuthRecordKey).(*models.Record)
	if record == nil {
		return nil, errors.New("The user record is not on the context.")
	}
	return record, nil
}

func Admin(c echo.Context) (*models.Admin, error) {
	admin, _ := c.Get(apis.ContextAdminKey).(*models.Admin)
	if admin == nil {
		return nil, errors.New("The admin is not on the context.")
	}
	return admin, nil
}
```

**Reply by ganigeorgiev** [reply]
> In your experience is that a bad thing? Are you actually using that middleware example somewhere like that, or is it purely an example?

It was just an example. Why it is a bad thing? It shouldn't matter whether you use one or separate cookies (especially if they have the same attributes). It shouldn't matter how it is stored in the local storage in this case (the 2 different keys are to ensure that when the SDK is used the admin auth state doesn't interfere with your application). Also keep in mind that the Admin UI will never use the cookie anyway. The above middleware is for your application.

---

> I also made some helpers like this below. I feel like that the context access stuff should be wrapped since it's such a common operation. Does the extra function call make a real difference in performance?

No, wrapping something in a function shouldn't have a noticeable impact on the performance.
Actually yesterday I was thinking adding helpers for that while working on https://github.com/pocketbase/pocketbase/discussions/943#discussioncomment-4167006, but at the end decided not to since the individual helpers don't really do much in this case - they just extract a value from the context and basically you are replacing a single line with another. Alternatively, we can create a custom echo.Context, where helpers like this will be part of the instance (eg. `c.AuthRecord()`), but the echo handler's declaration always expects the interface and users will have to cast it to the respective custom struct anyway (you can also check this discussion on adding support for generics in echo - https://github.com/labstack/echo/discussions/2000#discussioncomment-3959165).

**Reply by dschissler** [reply]
What about using this style to make sure that there aren't any collisions?  To be honest I've never dug into the details of how context keys are set, but I'm told that it's so.

```go
// Ensure no conflicts in naming.
type userKey string

const UserContextKey = userKey("auth-user")
```

**Reply by ganigeorgiev** [reply]
@dschissler I'm not sure that I understand what the above is supposed to do/fix. Could you elaborate a little more?

**Reply by dschissler** [reply]
Perhaps something like this?

```golang
user, err := pbtools.Context(c).AuthUser()
if err != nil {
  fmt.Println("You've got yourself a user")
}
```

I will be adapting my casbin route checker from Chi to Pocketbase/Echo.  I'd rather use a helper method with a real method name, and documentation than casting around everywhere.  Maybe it's not actually, but it somehow feels too brittle for me, especially around security.

**Reply by ganigeorgiev** [reply]
I'm not sure about this. For now I don't plan to change it because it adds an extra layer of abstraction without bringing too much value. Users will have to use the echo.Context store at some point or another if they want to pass local data between middlewares and hiding it only for the auth state seems redundant and just add another way of doing the same thing (not to mention that we'll have to add also support for a setter).

```go
user, err := pbtools.Context(c).AuthUser()
// vs
user, ok := c.Get(apis.ContextAuthRecordKey).(*models.Record)
```

**Reply by dschissler** [reply]
@ganigeorgiev Would it be too much to add two ways to access it?

```golang
// This allows the context to be stored in a variable and used multiple times
user, err := pbtools.Context(c).AuthUser()

// and also this is more direct.
user, err := pbsecurity.AuthUser(c)
/// or perhaps instead on pbtools
user, err := pbtools.ContextAuthUser(c)
```

**Reply by dschissler** [reply]
> I'm not sure about this. For now I don't plan to change it because it adds an extra layer of abstraction without bringing too much value. Users will have to use the echo.Context store at some point or another if they want to pass local data between middlewares and hiding it only for the auth state seems redundant and just add another way of doing the same thing (not to mention that we'll have to add also support for a setter).
> 
> ```go
> user, err := pbtools.Context(c).AuthUser()
> // vs
> user, ok := c.Get(apis.ContextAuthRecordKey).(*models.Record)
> ```

My thinking was so that it could be clean like this when doing multiple checks:

```golang
pbc := pbtools.Context(c)

user, err := pbc.AuthUser()
// ....
admin, err := pbc.AuthAdmin()
// ...
otherthing, err := pbc.OtherThing()
```

**Reply by ganigeorgiev** [reply]
@dschissler I'm usually ok with wrappers/aliases, but as mentioned previously, in this case this is just a single line that call `Get` with a constant.

There is nothing wrong with having your own PocketBase/echo helpers subpackage if you prefer the above access style.

**Reply by dschissler** [reply]
> @dschissler I'm not sure that I understand what the above is supposed to do/fix. Could you elaborate a little more?

The recommendation is in the Golang source for Context.WithValue
https://github.com/golang/go/blob/master/src/context/context.go#L574-L580

// The provided key must be comparable and **should not be of type
// string or any other built-in type to avoid collisions between
// packages using context.** Users of WithValue should define their own
// types for keys. To avoid allocating when assigning to an
// interface{}, context keys often have concrete type
// struct{}. Alternatively, exported context key variables' static
// type should be a pointer or interface.

So like this:

```go
// Ensure no conflicts in naming.
type userKey string

const UserContextKey = userKey("auth-user")
```

**Reply by ganigeorgiev** [reply]
@dschissler `echo.Context` is not the same as `context.Context`. `echo.Context.Get()` and `echo.Context.Set()` works with `string` keys.

**Reply by vanevalverdes** [reply]
Hi!

Exists any way to use this solution using pocketbase as framework in JS ?

Thanks

**Reply by ganigeorgiev** [reply]
@vanevalverdes It depends what is the desired outcome. For example how to load an auth context with cookies you can check https://github.com/pocketbase/pocketbase/discussions/3447. If that doesn't help, please file a new discussion with more details about your use case.

**Comment by dschissler** [comment]
@ganigeorgiev It's not a big deal, but why did you change the "user" methods to "record"?  For example, `app.OnUserAuthRequest` to `app.OnRecordAuthRequest`.

**Reply by ganigeorgiev** [reply]
Because in v0.8.0-r we no longer have `models.User` and all entries in a collection are `models.Record` (also note that you can have multiple auth collections, each with their own separate login configuration and custom fields).

**Comment by efim** [comment]
I was really interested in using Pocketbase for SSR go + templates + HTMX , especially having oauth avaialble for signup.
Here are results of my moderately successful experimentation:
repo: http://git.sunshine.industries/efim/go-ssr-pocketbase-oauth-attempt (with Readme)
and deployed site: https://go-ssr-oauth-attempt.sunshine.industries/

Admin page lives separately from the website pages.
There's ability to signup / authorize via oauth provider, that auth can be easily used to secure pages with existing or custom middlewares.
Middleware is used to set auth token into a secure cookie ( wouldn't be able to figure this out without code from @ganigeorgiev - thank you so much for the pocketbase, it looks awesome, hope to use it for some small project )
During page render go code can make queries into pocketbase DAO.
Also any existing pocketbase json web api's should just work, because same middleware would put auth token into correct place of context.

@dschissler 
I don't know much about Svetle, but if I understood your question, htmx deals with similar case in this way:
request which was triggered by htmx call and will need partial template to change only part of the page adds header `HX-Request: true`,
so server may choose to return render of only one specific template.
And when there is no such header - consider it a full page request by browser and return full page.
And for either of these requests user session will be added to the request context from the cookie.

**Comment by jonshipman** [comment]
I have a couple boilerplate examples:

https://github.com/jonshipman/sveltekit-pb-boilerplate
https://github.com/remix-run/examples/pull/409/commits/73dba077bb3dfbabf70b41ee7908972eee99f22e
